crates/corro-pg/src/codec.rs

```diff
@@ -0,0 +1,171 @@
+//! Reimpl of PgWireMessageServerCodec from <https://github.com/sunng87/pgwire/blob/master/src/tokio/server.rs>,
+//! the implementation was public, but was made (accidentally?) private in <https://github.com/sunng87/pgwire/pull/214>
+
+use pgwire::{
+    api::{self, ClientInfo},
+    error::PgWireError,
+    messages as msg,
+};
+use std::{collections::HashMap, io};
+use tokio_util::codec;
+
+pub struct Client {
+    pub socket_addr: std::net::SocketAddr,
+    pub is_secure: bool,
+    pub protocol_version: msg::ProtocolVersion,
+    pub pid_secret_key: (i32, msg::startup::SecretKey),
+    pub state: api::PgWireConnectionState,
+    pub transaction_status: msg::response::TransactionStatus,
+    pub metadata: HashMap<String, String>,
+}
+
+impl Client {
+    pub fn new(socket_addr: std::net::SocketAddr, is_secure: bool) -> Self {
+        Self {
+            socket_addr,
+            is_secure,
+            protocol_version: Default::default(),
+            pid_secret_key: Default::default(),
+            state: Default::default(),
+            transaction_status: msg::response::TransactionStatus::Idle,
+            metadata: Default::default(),
+        }
+    }
+}
+
+impl ClientInfo for Client {
+    #[inline]
+    fn socket_addr(&self) -> std::net::SocketAddr {
+        self.socket_addr
+    }
+
+    #[inline]
+    fn is_secure(&self) -> bool {
+        self.is_secure
+    }
+
+    #[inline]
+    fn state(&self) -> api::PgWireConnectionState {
+        self.state
+    }
+
+    #[inline]
+    fn set_state(&mut self, new_state: api::PgWireConnectionState) {
+        self.state = new_state;
+    }
+
+    #[inline]
+    fn transaction_status(&self) -> msg::response::TransactionStatus {
+        self.transaction_status
+    }
+
+    #[inline]
+    fn set_transaction_status(&mut self, new_status: msg::response::TransactionStatus) {
+        self.transaction_status = new_status;
+    }
+
+    #[inline]
+    fn metadata(&self) -> &HashMap<String, String> {
+        &self.metadata
+    }
+
+    #[inline]
+    fn metadata_mut(&mut self) -> &mut HashMap<String, String> {
+        &mut self.metadata
+    }
+
+    #[inline]
+    fn pid_and_secret_key(&self) -> (i32, msg::startup::SecretKey) {
+        self.pid_secret_key.clone()
+    }
+
+    #[inline]
+    fn set_pid_and_secret_key(&mut self, pid: i32, secret_key: msg::startup::SecretKey) {
+        self.pid_secret_key = (pid, secret_key);
+    }
+
+    #[inline]
+    fn protocol_version(&self) -> msg::ProtocolVersion {
+        self.protocol_version
+    }
+
+    #[inline]
+    fn set_protocol_version(&mut self, version: msg::ProtocolVersion) {
+        self.protocol_version = version;
+    }
+
+    #[inline]
+    fn client_certificates<'a>(&self) -> Option<&[rustls::pki_types::CertificateDer<'a>]> {
+        None
+    }
+}
+
+pub struct PgWireMessageServerCodec {
```

Comment:
This type was (accidentally?) made private in a code reorganization, so I've copied it here to reduce the churn in the lib file.